### PR TITLE
Add create getters and setters code actions for class level variables

### LINKED ISSUE #32004: Support getters and setters generation for class level variables
**Description:**
Add language server support to generate getters and setters for class level variables.

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
Fixes #32004

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package org.ballerinalang.langserver.codeaction.providers;
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java:1]: Missing license header

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
+package org.ballerinalang.langserver.codeaction.providers;
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java:1]: Missing license header

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java [hunk 3]
```diff
@@ -448,6 +451,24 @@ public static List<TextEdit> getTypeGuardCodeActionEdits(String varName, Range r
...
+    }
+
+    public static List<TextEdit> addSettersCodeActionEdits(String varName, Range range, int offset, String typeName) {
+        Position startPos = range.getEnd();
+        Range newTextRange = new Range(startPos, startPos);
+        List<TextEdit> edits = new ArrayList<>();
+        String spaces = StringUtils.repeat(' ', offset);
+        edits.add(new TextEdit(newTextRange, generateSetterFunctionBodyText(varName, typeName, spaces)));
+        return edits;
+    }
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java:470]: Shall we get the repeating code parts to a common place?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (!context.currentSyntaxTree().isPresent()) {
+            return Collections.emptyList();
+        }
+
+        NonTerminalNode matchedNode = CommonUtil.findNode(new Range(context.cursorPosition(), context.cursorPosition()),
+                context.currentSyntaxTree().get());
+        String commandTitle = "Create a getter for " + ((ObjectFieldNode) matchedNode).fieldName().toString();
+        String fieldName = String.valueOf(((ObjectFieldNode) matchedNode).fieldName());
+        String typeName = String.valueOf(((ObjectFieldNode) matchedNode).typeName());
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java:49]: Here we are evaluating the ((ObjectFieldNode) matchedNode) multiple times

**Review by gayalkuruppu — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    public List<CodeAction> getNodeBasedCodeActions(CodeActionContext context,
+                                                    NodeBasedPositionDetails posDetails) {
+
+        if (!context.currentSyntaxTree().isPresent()) {
+            return Collections.emptyList();
+        }
+
+        NonTerminalNode matchedNode = CommonUtil.findNode(new Range(context.cursorPosition(), context.cursorPosition()),
+                context.currentSyntaxTree().get());
+        String commandTitle = "Create a getter for " + ((ObjectFieldNode) matchedNode).fieldName().toString();
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java:47]: Shall we add inverted commas to the fieldName? And we can add `"Create a getter for %s" to the `CommandConstants.java`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package org.ballerinalang.langserver.codeaction.providers;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java:17]: Updated in 85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java (hunk 1 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java:17]: Updated in 85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java (hunk 2 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java:17]: Updated in 85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java (hunk 4 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java]: Fixed in 85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java (hunk 5 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java]: Added in 85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java [hunk 7]
```diff
@@ -865,6 +889,27 @@ private static String generateIfElseText(String varName, String spaces, String p
...
+
+    private static String generateSetterFunctionBodyText(String varName, String typeName, String spaces) {
+
+        StringBuilder newTextBuilder = new StringBuilder();
+        String functionName = varName.substring(0, 1).toUpperCase(Locale.ROOT) + varName.substring(1);
+        newTextBuilder.append(String.format
+                (LINE_SEPARATOR + LINE_SEPARATOR + spaces + "public function set%s(%s %s) { " + LINE_SEPARATOR +
+                spaces + spaces + "self.%s = %s;" + LINE_SEPARATOR + "" + spaces + "}" ,
+                functionName, typeName, varName, varName, varName));
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java:923]: Shall we append the strings separately?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ballerinalang.langserver.codeaction.providers;
+
+import io.ballerina.compiler.syntax.tree.*;
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java]: Replace with single class imports

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ballerinalang.langserver.codeaction.providers;
+
+import io.ballerina.compiler.syntax.tree.*;
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java]: Replace with single class imports

**Review by TharushiJay — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java [hunk 10]
```diff
@@ -30,25 +30,7 @@
...
-import io.ballerina.compiler.syntax.tree.NonTerminalNode;
-import io.ballerina.compiler.syntax.tree.ObjectTypeDescriptorNode;
-import io.ballerina.compiler.syntax.tree.RecordTypeDescriptorNode;
-import io.ballerina.compiler.syntax.tree.ReturnTypeDescriptorNode;
-import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
-import io.ballerina.compiler.syntax.tree.SyntaxKind;
-import io.ballerina.compiler.syntax.tree.SyntaxTree;
-import io.ballerina.compiler.syntax.tree.TypeDefinitionNode;
-import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import io.ballerina.compiler.syntax.tree.*;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/commit/85613cc5bfa90a74b38fb3e48f237c79dde35307

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java (hunk 3 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java]: Updated in 9b8ee2e211c6ab47082d00b3379ae13676eb610d

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java [hunk 11]
```diff
@@ -865,6 +889,27 @@ private static String generateIfElseText(String varName, String spaces, String p
         return LINE_SEPARATOR + newTextBuilder.toString();
     }
 
+    private static String generateGetterFunctionBodyText(String varName, String typeName, String spaces) {
+        StringBuilder newTextBuilder = new StringBuilder();
+        String functionName = varName.substring(0, 1).toUpperCase(Locale.ROOT) + varName.substring(1);
+        newTextBuilder.append(String.format
+                (LINE_SEPARATOR + LINE_SEPARATOR + spaces + "public function get%s() returns %s{ " + LINE_SEPARATOR +
+                spaces + spaces + "return self.%s;" + LINE_SEPARATOR + spaces + "}" ,
+                functionName, typeName, varName));
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java]: Updated in 9b8ee2e211c6ab47082d00b3379ae13676eb610d

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java (hunk 7 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java]: Updated in 9b8ee2e211c6ab47082d00b3379ae13676eb610d

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java (hunk 8 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterSetterCodeAction.java]: Fixed in 9b8ee2e211c6ab47082d00b3379ae13676eb610d

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java (hunk 9 above)
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/SetterCodeAction.java]: Fixed in 9b8ee2e211c6ab47082d00b3379ae13676eb610d

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        if (CodeActionUtil.getObjectFieldNode(context, posDetails).isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        ObjectFieldNode objectFieldNode = CodeActionUtil.getObjectFieldNode(context, posDetails).get();
+        String commandTitle = String.format("Create a getter for '%s'", objectFieldNode.fieldName().toString());
+        String fieldName = String.valueOf(objectFieldNode.fieldName());
+        String typeName = String.valueOf(objectFieldNode.typeName());
+        String functionName = "get" + fieldName.substring(0, 1).toUpperCase(Locale.ROOT) +
+                fieldName.substring(1);
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java:60]: Some lines are repeating across the code action files. Shall we explore the possibility to refactor that?

**Review by gayalkuruppu — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java [hunk 13]
```diff
@@ -891,6 +925,119 @@ private static String generateIfElseText(String varName, String spaces, String p
...
+            startLine = ((ClassDefinitionNode) objectFieldNode.parent()).
+                    members().get(((ClassDefinitionNode) objectFieldNode.parent()).members().size() - 1).
+                    lineRange().endLine().line();
+            startOffset = ((ClassDefinitionNode) objectFieldNode.parent()).
+                    members().get(((ClassDefinitionNode) objectFieldNode.parent()).members().size() - 1).
+                    lineRange().endLine().offset();
+            textOffset = objectFieldNode.lineRange().startLine().offset();
+        } else {
+            startLine = initNode.get().lineRange().endLine().line();
+            startOffset = initNode.get().lineRange().endLine().offset();
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java]: Updated in c384162c33784cdf5671868bd07ace10093faca5

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        if (CodeActionUtil.isfunctionDefined(functionName, objectFieldNode)) {
+            return Collections.emptyList();
+        }
+
+        Optional<FunctionDefinitionNode> initNode = CodeActionUtil.getInitNode(objectFieldNode);
+        List<TextEdit> edits = CodeActionUtil.getGetterSetterCodeEdits(objectFieldNode, initNode, fieldName, typeName,
+                                                                       NAME);
+        return Collections.singletonList(createCodeAction(commandTitle, edits, context.fileUri()));
```
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/GetterCodeAction.java:68]: We have decided to make it a `source action`. @DamithaSenevirathne let's change the code action kind

**Review by gayalkuruppu — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]
LGTM! There is some duplicated logic we can refactor into an abstract class. @DamithaSenevirathne please create an issue to track that. For example, the getter/setter code action can reuse stuff from getter code action and setter code action.

**Pull request merged**